janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java

```diff
@@ -4110,6 +4121,80 @@ public void testTinkerPopOptimizationStrategies() {
+                           gts.V(sv[0]).out("knows").is(vs[50]),
+                           AdjacentVertexIsOptimizerStrategy.instance());
+        assertOptimization(gts.V(sv[0]).inE("knows").has(ImplicitKey.ADJACENT_ID.name(), vs[50]).outV(),
+                           gts.V(sv[0]).in("knows").is(vs[50]),
+                           AdjacentVertexIsOptimizerStrategy.instance());
+        assertOptimization(gts.V(sv[0]).bothE("knows").has(ImplicitKey.ADJACENT_ID.name(), vs[50]).otherV(),
+                           gts.V(sv[0]).both("knows").is(vs[50]),
+                           AdjacentVertexIsOptimizerStrategy.instance());
+
+        // AdjacentVertexHasIdOptimizer outE/inE/bothE
+        assertOptimization(gts.V(sv[0]).outE("knows").has(ImplicitKey.ADJACENT_ID.name(), vs[50]).inV(),
+                           gts.V(sv[0]).outE("knows").inV().hasId(vs[50]),
+                           AdjacentVertexHasIdOptimizerStrategy.instance());
+        assertOptimization(gts.V(sv[0]).inE("knows").has(ImplicitKey.ADJACENT_ID.name(), vs[50]).outV(),
+                           gts.V(sv[0]).inE("knows").outV().hasId(vs[50]),
+                           AdjacentVertexHasIdOptimizerStrategy.instance());
+        assertOptimization(gts.V(sv[0]).bothE("knows").has(ImplicitKey.ADJACENT_ID.name(), vs[50]).otherV(),
+                           gts.V(sv[0]).bothE("knows").otherV().hasId(vs[50]),
+                           AdjacentVertexHasIdOptimizerStrategy.instance());
+
+        // AdjacentVertexHasIdOptimizer out/in/both
+        assertOptimization(gts.V(sv[0]).outE("knows").has(ImplicitKey.ADJACENT_ID.name(), vs[50]).inV(),
+                           gts.V(sv[0]).out("knows").hasId(vs[50]),
+                           AdjacentVertexHasIdOptimizerStrategy.instance());
+        assertOptimization(gts.V(sv[0]).inE("knows").has(ImplicitKey.ADJACENT_ID.name(), vs[50]).outV(),
+                           gts.V(sv[0]).in("knows").hasId(vs[50]),
+                           AdjacentVertexHasIdOptimizerStrategy.instance());
+        assertOptimization(gts.V(sv[0]).bothE("knows").has(ImplicitKey.ADJACENT_ID.name(), vs[50]).otherV(),
+                           gts.V(sv[0]).both("knows").hasId(vs[50]),
+                           AdjacentVertexHasIdOptimizerStrategy.instance());
+
+        // neq should not be optimized
+        assertOptimization(gts.V(sv[0]).in().hasId(P.neq(vs[50])),
+                           gts.V(sv[0]).in().hasId(P.neq(vs[50])),
+                           AdjacentVertexHasIdOptimizerStrategy.instance());
+
+        int[] loop1 = {0}; // repeat starts from vertex with id 0 and goes in to the sv[0] vertex then loops back out to the vertex with the next id
+        int[] loop2 = {0};
+        GraphTraversal t1 = gts.V(vs[0], vs[1], vs[2])
+                           .repeat(__.inE("knows")
+                                       .has("~adjacent", sv[0].id())
```

Comment:
Why don't you use `ImplicitKey.ADJACENT_ID` here but use `~adjacent`?

Reply:
This was just a result of frequent deleting and re-writing parts of the query in order to find the mentioned TinkerPop bug. At some point typing `"~adjacent"` was much quicker than typing `ImplicitKey.ADJACENT_ID.name()`. I will change this to match the style of the other tests